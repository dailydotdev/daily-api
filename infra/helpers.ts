import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
