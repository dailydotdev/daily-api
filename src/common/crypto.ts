import { webcrypto as crypto } from 'node:crypto';

const bufToBase64 = (x: ArrayBuffer) => Buffer.from(x).toString('base64');

export const encrypt = async (
  input: string,
  key: string,
  algorithmName = 'AES-CBC',
  algorithmLength = 256,
  ivLength = 16,
): Promise<string> => {
  const keyObject = await crypto.subtle.importKey(
    'raw',
    Buffer.from(key, 'utf-8'),
    {
      name: algorithmName,
      length: algorithmLength,
    },
    true,
    ['encrypt', 'decrypt'],
  );
  const iv = crypto.getRandomValues(new Uint8Array(ivLength));
  const encrypted = await crypto.subtle.encrypt(
    {
      name: algorithmName,
      iv,
    },
    keyObject,
    new TextEncoder().encode(input),
  );
  return bufToBase64(iv) + ':' + bufToBase64(encrypted);
};
