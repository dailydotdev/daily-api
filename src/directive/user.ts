import { DataSource } from 'typeorm';
import { FastifyLoggerInstance } from 'fastify';
import {
  Alerts,
  ArticlePost,
  Bookmark,
  BookmarkList,
  Comment,
  DevCard,
  Feed,
  Post,
  PostReport,
  Settings,
  SourceDisplay,
  SourceRequest,
  User,
  View,
} from '../entity';

export const deleteUser = async (
  con: DataSource,
  logger: FastifyLoggerInstance,
  userId: string,
  messageId?: string,
) => {
  try {
    await con.transaction(async (entityManager): Promise<void> => {
      await entityManager.getRepository(View).delete({ userId });
      await entityManager.getRepository(Alerts).delete({ userId });
      await entityManager.getRepository(BookmarkList).delete({ userId });
      await entityManager.getRepository(Bookmark).delete({ userId });
      await entityManager.getRepository(Comment).update(
        { userId },
        {
          userId: '404',
        },
      );
      await entityManager.getRepository(Comment).delete({ userId });
      await entityManager.getRepository(DevCard).delete({ userId });
      await entityManager.getRepository(Feed).delete({ userId });
      await entityManager.getRepository(PostReport).delete({ userId });
      await entityManager.getRepository(Settings).delete({ userId });
      await entityManager.getRepository(SourceDisplay).delete({ userId });
      await entityManager.getRepository(SourceRequest).delete({ userId });
      await entityManager
        .getRepository(ArticlePost)
        .update({ authorId: userId }, { authorId: null });
      // Manually set shared post to 404 dummy user
      await entityManager
        .getRepository(Post)
        .update({ authorId: userId }, { authorId: '404' });
      await entityManager
        .getRepository(Post)
        .update({ scoutId: userId }, { scoutId: null });
      await entityManager.getRepository(User).delete(userId);
    });
    if (logger) {
      logger.info(
        {
          userId,
          messageId,
        },
        'deleted user',
      );
    }
  } catch (err) {
    if (logger) {
      logger.error(
        {
          userId,
          messageId,
          err,
        },
        'failed to delete user',
      );
    }
    throw err;
  }
};
