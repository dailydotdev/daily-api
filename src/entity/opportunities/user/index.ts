export * from './OpportunityUser';
export * from './OpportunityUserRecruiter';
