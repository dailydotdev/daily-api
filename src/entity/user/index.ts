export * from './User';
export * from './UserPost';
export * from './UserState';
export * from './UserAction';
export * from './UserStreak';
export * from './UserPersonalizedDigest';
export * from './UserMarketingCta';
