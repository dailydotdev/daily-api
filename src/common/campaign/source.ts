import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
