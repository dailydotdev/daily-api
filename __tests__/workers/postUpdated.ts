import { expectSuccessfulBackground, saveFixtures } from '../helpers';
import worker from '../../src/workers/postUpdated';
import {
  ArticlePost,
  COMMUNITY_PICKS_SOURCE,
  Keyword,
  Post,
  PostKeyword,
  PostOrigin,
  PostType,
  SharePost,
  Source,
  Submission,
  SubmissionStatus,
  User,
} from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../../src/db';
import { randomUUID } from 'crypto';
import { usersFixture } from '../fixture/user';
import { SubmissionFailErrorMessage } from '../../src/errors';

let con: DataSource;

beforeAll(async () => {
  con = await createOrGetConnection();
});

beforeEach(async () => {
  jest.resetAllMocks();
  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, ArticlePost, [
    {
      id: 'p1',
      shortId: 'p1',
      url: 'http://p1.com',
      score: 0,
      metadataChangedAt: new Date('01-05-2020 12:00:00'),
      sourceId: 'a',
      visible: false,
      createdAt: new Date('01-05-2020 12:00:00'),
      origin: PostOrigin.Squad,
    },
  ]);
  await saveFixtures(con, ArticlePost, [
    {
      id: 'p2',
      shortId: 'p2',
      title: 'freeform post',
      score: 0,
      metadataChangedAt: new Date('01-05-2020 12:00:00'),
      sourceId: 'squad',
      visible: true,
      createdAt: new Date('01-05-2020 12:00:00'),
      type: PostType.Freeform,
      origin: PostOrigin.UserGenerated,
    },
  ]);
});

const createDefaultSubmission = async (id: string = randomUUID()) => {
  const repo = con.getRepository(Submission);
  await repo.save(
    repo.create({ id, url: 'http://sample.article/test', userId: '1' }),
  );
};

const createDefaultUser = async () => {
  await con
    .getRepository(User)
    .save({ ...usersFixture[0], twitter: 'leeTwitter' });
};

const createDefaultKeywords = async () => {
  const repo = con.getRepository(Keyword);
  await repo.insert({
    value: 'mongodb',
    status: 'allow',
  });
  await repo.insert({
    value: 'alpinejs',
    status: 'allow',
  });
  await repo.insert({
    value: 'ab-testing',
    status: 'allow',
  });
  await repo.insert({
    value: 'alpine',
    status: 'synonym',
    synonym: 'alpinejs',
  });
  await repo.insert({
    value: 'a-b-testing',
    status: 'synonym',
    synonym: 'ab-testing',
  });
};

const createSharedPost = async (id = 'sp1') => {
  const post = await con.getRepository(Post).findOneBy({ id: 'p1' });
  await con.getRepository(SharePost).save({
    ...post,
    id,
    shortId: `short-${id}`,
    sharedPostId: 'p1',
    type: PostType.Share,
    visible: false,
  });
};

it('should not update if the database updated date is newer', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-1990 12:00:00'),
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2020-01-05T12:00:00.000Z'));
});

it('should not update if the post is not a squad origin', async () => {
  await con
    .getRepository(ArticlePost)
    .update({ id: 'p1' }, { origin: PostOrigin.CommunityPicks });
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-1990 12:00:00'),
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2020-01-05T12:00:00.000Z'));
});

it(`should not update if the post object doesn't have ID`, async () => {
  await con
    .getRepository(ArticlePost)
    .update({ id: 'p1' }, { origin: PostOrigin.CommunityPicks });
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    updated_at: new Date('01-05-1990 12:00:00'),
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2020-01-05T12:00:00.000Z'));
});

it('should update the post and keep it invisible if title is missing', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-2023 12:00:00'),
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.visible).toEqual(false);
  expect(post.flags.visible).toEqual(false);
});

it('should update the post and make it visible if title is available', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-2023 12:00:00'),
    title: 'test',
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.visible).toEqual(true);
  expect(post.flags.visible).toEqual(true);
  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.title).toEqual('test');
});

it('should update all the post related shared posts to visible', async () => {
  await createSharedPost();
  await createSharedPost('sp2');
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-2023 12:00:00'),
    title: 'test',
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.visible).toEqual(true);
  expect(post.flags.visible).toEqual(true);
  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.title).toEqual('test');
  const sharedPost = await con
    .getRepository(SharePost)
    .findOneBy({ id: 'sp1' });
  expect(sharedPost.visible).toEqual(true);
  expect(sharedPost.flags.visible).toEqual(true);
  expect(sharedPost.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  const sharedPost2 = await con
    .getRepository(SharePost)
    .findOneBy({ id: 'sp2' });
  expect(sharedPost2?.visible).toEqual(true);
  expect(sharedPost2?.flags.visible).toEqual(true);
  expect(sharedPost2?.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
});

it('should update post and not modify keywords', async () => {
  await createDefaultUser();
  await createDefaultKeywords();
  await con
    .getRepository(Post)
    .update({ id: 'p1' }, { tagsStr: 'mongodb, alpinejs' });
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    title: 'New title',
    extra: {
      keywords: ['mongodb', 'alpinejs'],
    },
  });
  const post = await con.getRepository(Post).findOneBy({ id: 'p1' });
  expect(post.title).toEqual('New title');
  expect(post.tagsStr).toEqual('mongodb,alpinejs');
  const postKeywords = await con.getRepository(PostKeyword).find({
    where: {
      postId: 'p1',
    },
  });
  expect(postKeywords.length).toEqual(2);
});

it('should update post and modify keywords', async () => {
  await createDefaultUser();
  await createDefaultKeywords();
  await con
    .getRepository(Post)
    .update({ id: 'p1' }, { tagsStr: 'mongodb, alpinejs' });
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    title: 'New title',
    extra: {
      keywords: ['mongodb', 'ab-testing'],
    },
  });
  const post = await con.getRepository(Post).findOneBy({ id: 'p1' });
  expect(post.title).toEqual('New title');
  expect(post.tagsStr).toEqual('mongodb,ab-testing');
  const postKeywords = await con.getRepository(PostKeyword).find({
    where: {
      postId: 'p1',
    },
  });
  expect(postKeywords.length).toEqual(2);
});

it('should update freeform post and only modify allowed fields', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p2',
    updated_at: new Date('01-05-2023 12:00:00'),
    title: 'test',
    url: 'https://test.com',
    extra: {
      site_twitter: 'text',
      canonical_url: 'https://test.com/canon',
      content_curation: ['news', 'story', 'release'],
    },
    content_type: PostType.Freeform,
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p2' });
  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.visible).toEqual(true);
  expect(post.flags.visible).toEqual(true);
  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.contentCuration).toEqual(['news', 'story', 'release']);
  expect(post.siteTwitter).toEqual('text');
  expect(post.yggdrasilId).toEqual('f99a445f-e2fb-48e8-959c-e02a17f5e816');

  expect(post.title).toEqual('freeform post');
  expect(post.canonicalUrl).toBeNull();
  expect(post.url).toBeNull();
});

it('should save keywords without special characters', async () => {
  await createDefaultUser();
  await con
    .createQueryBuilder()
    .insert()
    .into(Keyword)
    .values([
      { value: 'abc', status: 'allow' },
      { value: 'ab', status: 'allow' },
      { value: 'a1-b2', status: 'allow' },
      { value: 'a_1.net', status: 'allow' },
      { value: '#c', status: 'allow' },
      { value: 'a-b', status: 'allow' },
      { value: '__', status: 'allow' },
    ])
    .execute();

  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    title: 'New title',
    extra: {
      keywords: [
        'abc',
        'a b ',
        'a1-b2',
        "'a1-b2'",
        'a_1.net',
        '#c',
        'a-b?',
        '_ãƒ„_',
        '?',
        "a'b",
      ],
    },
  });
  const post = await con.getRepository(Post).findOneBy({ id: 'p1' });
  expect(post.title).toEqual('New title');
  expect(post.tagsStr).toEqual('abc,ab,a1-b2,a_1.net,#c,a-b,__');
});

it('should save a new post with the relevant content curation', async () => {
  await createDefaultKeywords();
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    post_id: 'p1',
    updated_at: new Date('01-05-2023 12:00:00'),
    title: 'test',
    extra: {
      content_curation: ['news', 'story', 'release'],
    },
  });
  const post = await con.getRepository(ArticlePost).findOneBy({ id: 'p1' });
  expect(post.metadataChangedAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.visible).toEqual(true);
  expect(post.flags.visible).toEqual(true);
  expect(post.visibleAt).toEqual(new Date('2023-01-05T12:00:00.000Z'));
  expect(post.title).toEqual('test');
  expect(post.contentCuration).toEqual(['news', 'story', 'release']);
});

it('should save a new post with basic information', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    order: 0,
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2]).toMatchSnapshot({
    visible: true,
    visibleAt: expect.any(Date),
    createdAt: expect.any(Date),
    metadataChangedAt: expect.any(Date),
    score: expect.any(Number),
    id: expect.any(String),
    shortId: expect.any(String),
    contentCuration: expect.any(Array),
    sourceId: 'a',
    title: 'Title',
    showOnFeed: true,
  });
});

it('should set show on feed to true when order is missing', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].showOnFeed).toEqual(true);
});

it('should save a new post with showOnFeed information', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    order: 1,
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].showOnFeed).toEqual(false);
});

it('should save a new post with content curation', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    extra: {
      content_curation: ['news', 'story', 'release'],
    },
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].contentCuration).toStrictEqual(['news', 'story', 'release']);
});

it('save a post as public if source is public', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].private).toEqual(false);
  expect(posts[2].flags.private).toEqual(false);
});

it('save a post as private if source is private', async () => {
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'p',
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].private).toBe(true);
  expect(posts[2].flags.private).toBe(true);
});

it('should save a new post with the relevant scout id and update submission', async () => {
  const uuid = randomUUID();
  await saveFixtures(con, Source, [
    {
      id: COMMUNITY_PICKS_SOURCE,
      name: 'Community recommendations',
      image: 'sample.image.com',
    },
  ]);
  await createDefaultUser();
  await createDefaultSubmission(uuid);
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: COMMUNITY_PICKS_SOURCE,
    submission_id: uuid,
  });

  const post = await con
    .getRepository(ArticlePost)
    .findOneBy({ url: 'https://post.com' });
  expect(post.scoutId).toEqual('1');
  const submissions = await con.getRepository(Submission).find();
  const [submission] = submissions;
  expect(submissions.length).toEqual(1);
  expect(submission.id).toEqual(uuid);
  expect(submission.status).toEqual(SubmissionStatus.Accepted);
});

it('should save a new post with the relevant keywords', async () => {
  const uuid = randomUUID();
  await saveFixtures(con, Source, [
    {
      id: COMMUNITY_PICKS_SOURCE,
      name: 'Community recommendations',
      image: 'sample.image.com',
    },
  ]);
  await createDefaultUser();
  await createDefaultSubmission(uuid);
  await createDefaultKeywords();
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: `https://post.com/${uuid}`,
    source_id: COMMUNITY_PICKS_SOURCE,
    submission_id: uuid,
    extra: {
      keywords: ['alpine', 'a-b-testing', 'mongodb'],
    },
  });
  const posts = await con.getRepository(Post).find();
  expect(posts.length).toEqual(3);
  expect(posts[2].scoutId).toEqual('1');
  expect(posts[2].tagsStr).toEqual('mongodb,alpinejs,ab-testing');
  const keywords = await con.getRepository(Keyword).find({
    where: {
      value: 'alpine',
    },
  });
  // since I am adding a post which has `alpine`
  // as a tag, occurences of `alpine` in the db
  // should increase from 1 to 2
  expect(keywords[0].occurrences).toEqual(2);
});

it('should not accept post with same author and scout', async () => {
  const uuid = randomUUID();
  await createDefaultUser();
  await createDefaultSubmission(uuid);
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    submission_id: uuid,
    extra: {
      creator_twitter: 'leeTwitter',
    },
  });
  const submissions = await con.getRepository(Submission).find();
  const [submission] = submissions;
  expect(submissions.length).toEqual(1);
  expect(submission.id).toEqual(uuid);
  expect(submission.status).toEqual(SubmissionStatus.Rejected);
  expect(submission.reason).toEqual('SCOUT_IS_AUTHOR');
});

it('should update submission to rejected', async () => {
  const uuid = randomUUID();
  await createDefaultUser();
  await createDefaultSubmission(uuid);
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    submission_id: uuid,
    reject_reason: SubmissionFailErrorMessage.PAYWALL,
  });
  const submissions = await con.getRepository(Submission).find();
  const [submission] = submissions;
  expect(submissions.length).toEqual(1);
  expect(submission.id).toEqual(uuid);
  expect(submission.status).toEqual(SubmissionStatus.Rejected);
});

it('should not update already approved post', async () => {
  const uuid = randomUUID();
  await createDefaultUser();
  const repo = con.getRepository(Submission);
  await repo.save({
    id: uuid,
    url: 'http://sample.article/test',
    userId: '1',
    status: SubmissionStatus.Accepted,
  });
  await expectSuccessfulBackground(worker, {
    id: 'f99a445f-e2fb-48e8-959c-e02a17f5e816',
    title: 'Title',
    url: 'https://post.com',
    source_id: 'a',
    submission_id: uuid,
    reject_reason: SubmissionFailErrorMessage.PAYWALL,
  });
  const submissions = await con.getRepository(Submission).find();
  const [submission] = submissions;
  expect(submissions.length).toEqual(1);
  expect(submission.id).toEqual(uuid);
  expect(submission.status).toEqual(SubmissionStatus.Accepted);
});
