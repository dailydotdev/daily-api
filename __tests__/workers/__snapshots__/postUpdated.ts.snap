// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should save a new post with basic information 1`] = `
Object {
  "authorId": null,
  "banned": false,
  "canonicalUrl": "https://post.com",
  "comments": 0,
  "contentCuration": Any<Array>,
  "contentMeta": Any<Object>,
  "contentQuality": Any<Object>,
  "createdAt": Any<Date>,
  "creatorTwitter": null,
  "deleted": false,
  "description": null,
  "discussionScore": null,
  "downvotes": 0,
  "flags": Object {
    "private": false,
    "sentAnalyticsReport": true,
    "showOnFeed": true,
    "visible": true,
  },
  "id": Any<String>,
  "image": null,
  "language": "en",
  "lastTrending": null,
  "metadataChangedAt": Any<Date>,
  "origin": "crawler",
  "pinnedAt": null,
  "placeholder": null,
  "private": false,
  "publishedAt": null,
  "ratio": null,
  "readTime": null,
  "score": Any<Number>,
  "scoutId": null,
  "sentAnalyticsReport": true,
  "shortId": Any<String>,
  "showOnFeed": true,
  "siteTwitter": null,
  "slug": Any<String>,
  "sourceId": "a",
  "summary": null,
  "tagsStr": null,
  "title": "Title",
  "titleHtml": null,
  "toc": null,
  "trending": null,
  "type": "article",
  "upvotes": 0,
  "url": "https://post.com",
  "views": 0,
  "viewsThreshold": 0,
  "visible": true,
  "visibleAt": Any<Date>,
  "yggdrasilId": "f99a445f-e2fb-48e8-959c-e02a17f5e816",
}
`;
